In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

Simfin is where we will be downloading our fundamentals data from. Instead of downloading a CSV file, we will use the API key to download directly.

In [2]:
import simfin as sf

# Set your SimFin + API-key for downloading data.
sf.set_api_key('nYdjeL237mt7QaQ4OZTaPa4Xn3YECrTH')

# Set the local directory where data-files are stored.
# The directory will be created if it does not already exist.
sf.set_data_dir('~/simfin_data/')

# Download the data from the SimFin server and load into a Pandas DataFrame.
# We will be getting 3 sets of financial statements (income statement, balance sheet, cashflow statement)
df_income = sf.load_income(variant='quarterly', market='us')
df_balance = sf.load_balance(variant='quarterly', market='us')
df_cashflow = sf.load_cashflow(variant='quarterly', market='us')

# Print the first rows of the data.
df_income.head()

Dataset "us-income-quarterly" on disk (5 days old).
- Loading from disk ... Done!
Dataset "us-balance-quarterly" on disk (5 days old).
- Loading from disk ... Done!
Dataset "us-cashflow-quarterly" on disk (5 days old).
- Loading from disk ... Done!


SimFinId Currency  Fiscal Year Fiscal Period Publish Date  \
Ticker Report Date                                                              
A      2016-10-31      45846      USD         2016            Q4   2016-12-20   
       2017-01-31      45846      USD         2017            Q1   2017-03-08   
       2017-04-30      45846      USD         2017            Q2   2017-06-06   
       2017-07-31      45846      USD         2017            Q3   2017-09-06   
       2017-10-31      45846      USD         2017            Q4   2017-12-21   

                   Restated Date  Shares (Basic)  Shares (Diluted)  \
Ticker Report Date                                                   
A      2016-10-31     2018-12-20     324000000.0       328000000.0   
       2017-01-31     2018-03-06     322000000.0       326000000.0   
       2017-04-30     2018-05-31     321000000.0       325000000.0   
       2017-07-31     2018-08-30     321000000.0       326000000.0   
       2017-10-31     2018-12-20     324000000.0       327000000.0   

                         Revenue  Cost of Revenue  ...  \
Ticker Report Date                                 ...   
A      2016-10-31   1.111000e+09     -523000000.0  ...   
       2017-01-31   1.067000e+09     -493000000.0  ...   
       2017-04-30   1.102000e+09     -510000000.0  ...   
       2017-07-31   1.114000e+09     -518000000.0  ...   
       2017-10-31   1.189000e+09     -542000000.0  ...   

                    Non-Operating Income (Loss)  Interest Expense, Net  \
Ticker Report Date                                                       
A      2016-10-31                   -32000000.0            -16000000.0   
       2017-01-31                   -13000000.0            -16000000.0   
       2017-04-30                   -10000000.0            -15000000.0   
       2017-07-31                    -8000000.0            -13000000.0   
       2017-10-31                    -7000000.0            -13000000.0   

                    Pretax Income (Loss), Adj.  Abnormal Gains (Losses)  \
Ticker Report Date                                                        
A      2016-10-31                    151000000                      NaN   
       2017-01-31                    193000000                      NaN   
       2017-04-30                    191000000                      NaN   
       2017-07-31                    193000000                      NaN   
       2017-10-31                    226000000                      NaN   

                    Pretax Income (Loss)  Income Tax (Expense) Benefit, Net  \
Ticker Report Date                                                            
A      2016-10-31              151000000                        -25000000.0   
       2017-01-31              193000000                        -25000000.0   
       2017-04-30              191000000                        -27000000.0   
       2017-07-31              193000000                        -18000000.0   
       2017-10-31              226000000                        -49000000.0   

                    Income (Loss) from Continuing Operations  \
Ticker Report Date                                             
A      2016-10-31                                  126000000   
       2017-01-31                                  168000000   
       2017-04-30                                  164000000   
       2017-07-31                                  175000000   
       2017-10-31                                  177000000   

                    Net Extraordinary Gains (Losses)  Net Income  \
Ticker Report Date                                                 
A      2016-10-31                                NaN   126000000   
       2017-01-31                                NaN   168000000   
       2017-04-30                                NaN   164000000   
       2017-07-31                                NaN   175000000   
       2017-10-31                                NaN   177000000   

                    Net Incom

In [3]:
df_balance.head()

SimFinId Currency  Fiscal Year Fiscal Period Publish Date  \
Ticker Report Date                                                              
A      2016-10-31      45846      USD         2016            Q4   2016-12-20   
       2017-01-31      45846      USD         2017            Q1   2017-03-08   
       2017-04-30      45846      USD         2017            Q2   2017-06-06   
       2017-07-31      45846      USD         2017            Q3   2017-09-06   
       2017-10-31      45846      USD         2017            Q4   2017-12-21   

                   Restated Date  Shares (Basic)  Shares (Diluted)  \
Ticker Report Date                                                   
A      2016-10-31     2017-12-21     324000000.0       328000000.0   
       2017-01-31     2017-03-08     322000000.0       326000000.0   
       2017-04-30     2017-06-06     321000000.0       325000000.0   
       2017-07-31     2017-09-06     321000000.0       326000000.0   
       2017-10-31     2018-12-20     324000000.0       327000000.0   

                    Cash, Cash Equivalents & Short Term Investments  \
Ticker Report Date                                                    
A      2016-10-31                                      2.289000e+09   
       2017-01-31                                      2.241000e+09   
       2017-04-30                                      2.389000e+09   
       2017-07-31                                      2.563000e+09   
       2017-10-31                                      2.678000e+09   

                    Accounts & Notes Receivable  ...  Short Term Debt  \
Ticker Report Date                               ...                    
A      2016-10-31                   631000000.0  ...              0.0   
       2017-01-31                   653000000.0  ...      190000000.0   
       2017-04-30                   677000000.0  ...      241000000.0   
       2017-07-31                   678000000.0  ...      280000000.0   
       2017-10-31                   724000000.0  ...      210000000.0   

                    Total Current Liabilities  Long Term Debt  \
Ticker Report Date                                              
A      2016-10-31                   945000000    1.904000e+09   
       2017-01-31                  1089000000    1.802000e+09   
       2017-04-30                  1187000000    1.802000e+09   
       2017-07-31                  1241000000    1.801000e+09   
       2017-10-31                  1263000000    1.801000e+09   

                    Total Noncurrent Liabilities  Total Liabilities  \
Ticker Report Date                                                    
A      2016-10-31                   2.603000e+09         3548000000   
       2017-01-31                   2.483000e+09         3572000000   
       2017-04-30                   2.454000e+09         3641000000   
       2017-07-31                   2.409000e+09         3650000000   
       2017-10-31                   2.328000e+09         3591000000   

                    Share Capital & Additional Paid-In Capital  \
Ticker Report Date                                               
A      2016-10-31                                 9.165000e+09   
       2017-01-31                                 5.239000e+09   
       2017-04-30                                 5.242000e+09   
       2017-07-31                                 5.285000e+09   
       2017-10-31                                 5.303000e+09   

                    Treasury Stock  Retained Earnings  Total Equity  \
Ticker Report Date                                                    
A      2016-10-31    -1.050800e+10       6.089000e+09  4.246000e+09   
       2017-01-31     0.000000e+00      -4.530000e+08  4.300000e+09   
       2017-04-30     0.000000e+00      -3.930000e+08  4.375000e+09   
       2017-07-31     0.000000e+00      -2.600000e+08  4.611000e+09   
       2017-10-31              NaN      -1.260000e+08  4.835000e+09   

                    Total Liabilitie

In [4]:
df_cashflow.head()

SimFinId Currency  Fiscal Year Fiscal Period Publish Date  \
Ticker Report Date                                                              
A      2016-10-31      45846      USD         2016            Q4   2016-12-20   
       2017-01-31      45846      USD         2017            Q1   2017-03-08   
       2017-04-30      45846      USD         2017            Q2   2017-06-06   
       2017-07-31      45846      USD         2017            Q3   2017-09-06   
       2017-10-31      45846      USD         2017            Q4   2017-12-21   

                   Restated Date  Shares (Basic)  Shares (Diluted)  \
Ticker Report Date                                                   
A      2016-10-31     2018-12-20     324000000.0       328000000.0   
       2017-01-31     2018-03-06     322000000.0       326000000.0   
       2017-04-30     2018-05-31     321000000.0       325000000.0   
       2017-07-31     2018-08-30     321000000.0       326000000.0   
       2017-10-31     2018-12-20     324000000.0       327000000.0   

                    Net Income/Starting Line  Depreciation & Amortization  \
Ticker Report Date                                                          
A      2016-10-31                126000000.0                   56000000.0   
       2017-01-31                168000000.0                   55000000.0   
       2017-04-30                164000000.0                   54000000.0   
       2017-07-31                175000000.0                   51000000.0   
       2017-10-31                177000000.0                   52000000.0   

                    ...  Net Cash from Operating Activities  \
Ticker Report Date  ...                                       
A      2016-10-31   ...                         234000000.0   
       2017-01-31   ...                         116000000.0   
       2017-04-30   ...                         257000000.0   
       2017-07-31   ...                         228000000.0   
       2017-10-31   ...                         288000000.0   

                    Change in Fixed Assets & Intangibles  \
Ticker Report Date                                         
A      2016-10-31                            -52000000.0   
       2017-01-31                            -32000000.0   
       2017-04-30                            -43000000.0   
       2017-07-31                            -43000000.0   
       2017-10-31                            -58000000.0   

                    Net Change in Long Term Investment  \
Ticker Report Date                                       
A      2016-10-31                                  0.0   
       2017-01-31                                  NaN   
       2017-04-30                                  NaN   
       2017-07-31                                  NaN   
       2017-10-31                                  0.0   

                    Net Cash from Acquisitions & Divestitures  \
Ticker Report Date                                              
A      2016-10-31                                 -26000000.0   
       2017-01-31                                 -69000000.0   
       2017-04-30                                         0.0   
       2017-07-31                                 -57000000.0   
       2017-10-31                                         0.0   

                    Net Cash from Investing Activities  Dividends Paid  \
Ticker Report Date                                                       
A      2016-10-31                          -78000000.0     -38000000.0   
       2017-01-31                         -101000000.0     -42000000.0   
       2017-04-30                          -43000000.0     -43000000.0   
       2017-07-31                         -101000000.0     -42000000.0   
       2017-10-31                          -60000000.0     -43000000.0   

                    Cash from (Repayment of) Debt  \
Ticker Report Date                                  
A      2016-10-31                      27000000.0   
       2017-01-31

In [6]:
print('Income Statement CSV data is: ', df_income.shape)
print('Balance Sheet CSV data is: ', df_balance.shape)
print('Cash Flow CSV data is: ', df_cashflow.shape)

Income Statement CSV data is:  (36685, 26)
Balance Sheet CSV data is:  (36685, 28)
Cash Flow CSV data is:  (36685, 26)


Number of rows of data is consistent. There are altogether 36685 rows to use.

In [10]:
# Merge the data together
# Define the column features where merge takes place
list_to_merge_on = ['Ticker', 'SimFinId', 'Currency', 'Fiscal Year', 'Report Date', 'Publish Date']

# Merge the income statement and balance sheet first
merge1 = pd.merge(df_income, df_balance, on = list_to_merge_on, how = 'inner')

# Merge previous result with cashflow statement
df_merged = pd.merge(merge1, df_cashflow, on = list_to_merge_on, how = 'inner')

# Reset the index
df_merged.reset_index(inplace=True)

# Make sure that the dates are in correct format
df_merged["Report Date"] = pd.to_datetime(df_merged["Report Date"])
df_merged["Publish Date"] = pd.to_datetime(df_merged["Publish Date"])

print('Merged X data matrix shape is: ', df_merged.shape)

Merged X data matrix shape is:  (36685, 74)


In [11]:
df_merged.head()

Ticker Report Date  SimFinId Currency  Fiscal Year Fiscal Period_x  \
0      A  2016-10-31     45846      USD         2016              Q4   
1      A  2017-01-31     45846      USD         2017              Q1   
2      A  2017-04-30     45846      USD         2017              Q2   
3      A  2017-07-31     45846      USD         2017              Q3   
4      A  2017-10-31     45846      USD         2017              Q4   

  Publish Date Restated Date_x  Shares (Basic)_x  Shares (Diluted)_x  ...  \
0   2016-12-20      2018-12-20       324000000.0         328000000.0  ...   
1   2017-03-08      2018-03-06       322000000.0         326000000.0  ...   
2   2017-06-06      2018-05-31       321000000.0         325000000.0  ...   
3   2017-09-06      2018-08-30       321000000.0         326000000.0  ...   
4   2017-12-21      2018-12-20       324000000.0         327000000.0  ...   

   Net Cash from Operating Activities  Change in Fixed Assets & Intangibles  \
0                         234000000.0                           -52000000.0   
1                         116000000.0                           -32000000.0   
2                         257000000.0                           -43000000.0   
3                         228000000.0                           -43000000.0   
4                         288000000.0                           -58000000.0   

   Net Change in Long Term Investment  \
0                                 0.0   
1                                 NaN   
2                                 NaN   
3                                 NaN   
4                                 0.0   

   Net Cash from Acquisitions & Divestitures  \
0                                -26000000.0   
1                                -69000000.0   
2                                        0.0   
3                                -57000000.0   
4                                        0.0   

   Net Cash from Investing Activities  Dividends Paid  \
0                         -78000000.0     -38000000.0   
1                        -101000000.0     -42000000.0   
2                         -43000000.0     -43000000.0   
3                        -101000000.0     -42000000.0   
4                         -60000000.0     -43000000.0   

   Cash from (Repayment of) Debt  Cash from (Repurchase of) Equity  \
0                     27000000.0                       -43000000.0   
1                     89000000.0                       -93000000.0   
2                     52000000.0                       -75000000.0   
3                     39000000.0                        32000000.0   
4                    -70000000.0                         8000000.0   

   Net Cash from Financing Activities  Net Change in Cash  
0                         -56000000.0            90000000  
1                         -58000000.0           -48000000  
2                         -67000000.0           148000000  
3                          29000000.0           174000000  
4                        -106000000.0           115000000  

[5 rows x 74 columns]

In [18]:
df_prices = sf.load_shareprices(variant='daily', market='us')

Dataset "us-shareprices-daily" on disk (5 days old).
- Loading from disk ... Done!


In [13]:
df_prices.reset_index(inplace=True)

In [14]:
df_prices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3157862 entries, 0 to 3157861
Data columns (total 11 columns):
 #   Column              Dtype         
---  ------              -----         
 0   Ticker              object        
 1   Date                datetime64[ns]
 2   SimFinId            int64         
 3   Open                float64       
 4   Low                 float64       
 5   High                float64       
 6   Close               float64       
 7   Adj. Close          float64       
 8   Dividend            float64       
 9   Volume              int64         
 10  Shares Outstanding  float64       
dtypes: datetime64[ns](1), float64(7), int64(2), object(1)
memory usage: 265.0+ MB


In [15]:
df_prices.isnull().sum()

Ticker                      0
Date                        0
SimFinId                    0
Open                    21601
Low                     22052
High                    21166
Close                   21571
Adj. Close              21456
Dividend              3134252
Volume                      0
Shares Outstanding      44696
dtype: int64

In [40]:
# This is a function to get price data between two dates of a particular ticker
# return price and volume only

def getPrice (ticker, datestart, how_many_days, df=df_prices):
    windowDays = 5
    rows = df[
    (df["Date"].between(pd.to_datetime(datestart) + pd.Timedelta(days=how_many_days),\
                             pd.to_datetime(datestart) + pd.Timedelta(days=windowDays + how_many_days)))\
        & (df["Ticker"]==ticker)]
    
    if rows.empty:
        return [ticker, np.float("NaN"),\
        np.datetime64('NaT'),\
        np.float("NaN")]
    else:
        return [ticker, rows.iloc[0]["Open"],\
        rows.iloc[0]["Date"],\
        rows.iloc[0]["Volume"]*rows.iloc[0]["Open"]]

In [42]:
def getYPriceDataNearDate(ticker, date, modifier, dailySharePrices):
    windowDays=5
    rows = dailySharePrices[
    (dailySharePrices["Date"].between(pd.to_datetime(date)
    + pd.Timedelta(days=modifier),
    pd.to_datetime(date)
    + pd.Timedelta(days=windowDays
    +modifier)
    )
    ) & (dailySharePrices["Ticker"]==ticker)]
    if rows.empty:
        return [ticker, np.float("NaN"),\
        np.datetime64('NaT'),\
        np.float("NaN")]
    else:
        return [ticker, rows.iloc[0]["Open"],\
        rows.iloc[0]["Date"],\
        rows.iloc[0]["Volume"]*rows.iloc[0]["Open"]]

In [44]:
getYPriceDataNearDate('A', '2016-10-31', 2, df_prices)

KeyError: 'Date'

In [27]:
def getYPricesReportDateAndTargetDate(x, d, modifier=365):
    ''' Takes in all fundamental data X, all stock prices over time y,
    and modifier (days), and returns the stock price info for the
    data report date, as well as the stock price one year from that date
    (if modifier is left as modifier=365)
    '''
    # Preallocation list of list of 2
    # [(price at date) (price at date + modifier)]
    y = [[None]*8 for i in range(len(x))]
    whichDateCol='Publish Date'# or 'Report Date',
    # is the performance date from->to. Want this to be publish date.
    # Because of time lag between report date
    # (which can't be actioned on) and publish date
    # (data we can trade with)
    i=0
    for index in range(len(x)):
        y[i]=(getYPriceDataNearDate(x['Ticker'].iloc[index],
        x[whichDateCol].iloc[index],0,d)
        +getYPriceDataNearDate(x['Ticker'].iloc[index],
        x[whichDateCol].iloc[index],
        modifier, d))
        i=i+1
    return y